File: src/hermes_swerve_module/hermes_swerve_module/swerve_module.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64
from sensor_msgs.msg import JointState
from hermes_swerve_module.pid import PIDController
from geometry_msgs.msg import Twist

CSI = '\033['

def code_to_chars(code):
    return CSI + str(code) + 'm'
class AnsiCodes(object):
    def __init__(self):
        # the subclasses declare class attributes which are numbers.
        # Upon instantiation we define instance attributes, which are the same
        # as the class attributes but wrapped with the ANSI escape sequence
        for name in dir(self):
            if not name.startswith('_'):
                value = getattr(self, name)
                setattr(self, name, code_to_chars(value))
                
class AnsiFore(AnsiCodes):
    BLACK           = 30
    RED             = 31
    GREEN           = 32
    YELLOW          = 33
    BLUE            = 34
    MAGENTA         = 35
    CYAN            = 36
    WHITE           = 37
    RESET           = 39

Fore   = AnsiFore()

class Module(Node):
    def __init__(self, name):
        
        """
        For this, I've made a distinction between actual, target, and commanded values
        
        Actual: The actual value of the hardware
        Requesting: What the user wants to set the value to
        Commanding: What the controller is commanding the hardware to do
        
        """
        
        # Use the namespace parameter
        super().__init__(name)
        
        self.odom_frequency = 10.0
        
        self.warning_flag_for_dime = False
        self.module_name = self.declare_parameter('module_name', 'module_default').get_parameter_value().string_value
        self.mock_encoder_values = self.declare_parameter('mock_encoder_values', True).get_parameter_value().bool_value # For running without actual hardware
        self.invert_drive_motor = self.declare_parameter('invert_drive_motor', False).get_parameter_value().bool_value # For running without actual hardware
        # self.invert_drive_motor = True
        self.get_logger().info(f'Creating Module: {name}')
        
        # Print out that the invert_drive_motor parameter is set
        if self.invert_drive_motor:
            self.get_logger().info(f'{Fore.GREEN}Inverting Drive Motor for {self.module_name}{Fore.RESET}')
        
        # Publish Encoder values and limit switch state
        self.encoder_sub = self.create_publisher(JointState, f'/{self.module_name}/encoders', 10)
        self.limit_switch_sub = self.create_publisher(Float64, f'/{self.module_name}/limit_switch', 10)

        # Pub/Sub for target pivot angle
        # self.rqst_pivot_sub = self.create_subscription(Float64, f'/{self.module_name}/rqst_pivot_angle', self.set_rqst_pivot_angle, 10)
        self.pivot_position_pub = self.create_publisher(Float64, f'/{self.module_name}/pivot_position', 10)
        
        # Move all request commands to a common "Twist" command, where X is the speed and Z is the pivot angle
        self.command_sub = self.create_subscription(Twist, f'/{self.module_name}/command', self.command_callback, 10)
        
        # Pub/Sub for motor speed
        # self.drive_motor_sub = self.create_subscription(Float64, f'/{self.module_name}/rqst_wheel_speed', self.set_rqst_wheel_speed, 10)
        self.drive_motor_pub = self.create_publisher(Float64, f'/{self.module_name}/wheel_position', 10)
        self.speed_motor_pub = self.create_publisher(Float64, f'/{self.module_name}/wheel_speed', 10)
        
        # Create an error publisher for the pivot direction
        self.pivot_error_pub = self.create_publisher(Float64, f'/{self.module_name}/pivot_error', 10)
        
        # Publish the error at an interval (compute_pivot_error)
        self.error_timer = self.create_timer(
            1 / 2,  # Publish every 2 seconds to reduce CPU usage
            self.compute_pivot_error
        )
        self.pivot_error = 0.0  # Initialize pivot error

        # Internal state variables for module
        self.actual_wheel_speed = 0.0    # actual drive velocity from encoder
        self.rqst_wheel_speed = 0.0    # rqst drive velocity
        self.drive_wheel_position = 0.0
        self.target_wheel_position = 0.0
        

        self.rqst_pivot_angle = 0.0    # rqst pivot angle
        self.actual_pivot_angle = 0.0    # actual pivot angle from encoder
        
        
        self.commanded_wheel_speed = 0.0  # commanded drive velocity
        self.commanded_pivot_position = 0.0  # commanded pivot angle
        
        self.limit_switch_triggered = False
        
        # PID Controller for the drive
        self.drive_pid_controller = PIDController(p=0.1, i=0.4, d=0.01, target=self.rqst_wheel_speed, max_output=2, min_output=-2, logger=self.get_logger())

        # PID Controller for the pivot
        self.pivot_pid_controller = PIDController(p=0.1, i=0.1, d=0.01, target=self.rqst_pivot_angle, max_output=6.28, min_output=0.0, logger=self.get_logger())
        
        # Occassionally update the positions based on the differnentials
        self.timer = self.create_timer(
            1 / self.odom_frequency, self.update_position
        )
                
    def command_callback(self, msg):
        # Set the requested wheel speed and pivot angle based on the Twist message
        self.set_rqst_pivot_angle(msg.angular.z)    # Use the angular.z for pivot angle, add offset to match the simulation
        self.compute_pivot_error()                  # Compute the pivot error after setting the angle
        self.set_rqst_wheel_speed(msg.linear.x)     # Use the linear.x for wheel speed
        
    def compute_pivot_error(self):
        error = self.rqst_pivot_angle - self.actual_pivot_angle
        
        # Publish the pivot error for debugging purposes
        pivot_error_msg = Float64()
        pivot_error_msg.data = error
        self.pivot_error = error
        self.pivot_error_pub.publish(pivot_error_msg)
            

    def set_rqst_wheel_speed(self, speed : float):
        # Method to set drive speed
        if (self.pivot_error > 0.1):
            self.rqst_wheel_speed = self.rqst_wheel_speed * 0.95 # Reduce speed if the pivot error is too high to try and correct it
        else:
            self.rqst_wheel_speed = speed

    def set_rqst_pivot_angle(self, angle : float):
        # Method to set pivot angle
        self.rqst_pivot_angle = angle

    def update_encoder_values(self):
        # Callback to receive encoder feedback
        if not self.mock_encoder_values:
            # Get the actual encoder values
            raise NotImplementedError('Actual encoder values not implemented yet')
        else:
            # Mock encoder values for testing without actual hardware, this means that the encoder values match the commanded values
            self.actual_wheel_speed = self.commanded_wheel_speed
            self.actual_pivot_angle = self.commanded_pivot_position
            

    def update_limit_switch(self):
        # Callback to receive limit switch state
        if not self.mock_encoder_values:
            # Get the actual limit switch state
            raise NotImplementedError('Actual limit switch state not implemented yet')
        else:
            # Mock limit switch state for testing without actual hardware
            self.limit_switch_triggered = False

    def update_position(self):
        # log some values for me to see
        # self.get_logger().info(f'{self.module_name}: RPA: {self.rqst_pivot_angle} | AP: {self.actual_pivot_angle} | RWS: {self.rqst_wheel_speed} | AWS: {self.actual_wheel_speed}')
        
        self.update_encoder_values()
        self.update_limit_switch()
        
        # -------------------------- Driving Wheel --------------------------------------
        drive_output = self.drive_pid_controller.compute_move(self.rqst_wheel_speed, self.actual_wheel_speed)
        
        # Publish the wheel speed
        speed_msg = Float64()
        speed_msg.data = self.actual_wheel_speed
        self.speed_motor_pub.publish(speed_msg)
        
        self.drive_wheel_position += drive_output if not self.invert_drive_motor else -drive_output
        # self.get_logger().info(f'{self.module_name}: Drive Wheel Position: {self.drive_wheel_position} | Requested Wheel Speed: {self.rqst_wheel_speed} | Actual Wheel Speed: {self.actual_wheel_speed}')
        
        drive_msg = Float64()
        drive_msg.data = self.drive_wheel_position
        
        # consider self.self.invert_drive_motor
        self.drive_motor_pub.publish(drive_msg)
        
        # -------------------------- Pivot Wheel --------------------------------------
        modified_rqst_pivot_angle = self.rqst_pivot_angle
        pivot_output = self.pivot_pid_controller.compute_move(modified_rqst_pivot_angle, self.actual_pivot_angle)
        pivot_msg = Float64()
        pivot_msg.data = pivot_output # add the offset to the pivot angle to keep it synced with the sim 
        self.pivot_position_pub.publish(pivot_msg)
        
        # if we are mocking the hardware, then the actual values are the commanded values
        if self.mock_encoder_values:
            self.commanded_wheel_speed = drive_output
            self.commanded_pivot_position = pivot_output
            
def main(args=None):
    rclpy.init(args=args)

    # Instantiate the module with the specified namespace
    module = Module(name='hermes_swerve_module')

    # Spin to keep the node active
    try:
        rclpy.spin(module)
    except KeyboardInterrupt:
        pass

    # Shutdown the node
    module.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/hermes_robot_cpp/src/swerve_module.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64.hpp"
#include "sensor_msgs/msg/joint_state.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "hermes_robot_cpp/pid_controller.hpp"

using std::placeholders::_1;

/*

Note that, for this module, linear.x and angular.z are the wheel speed and pivot angle respectively

*/

// create a set of enums to track state of the module
enum class ModuleState
{
    IDLE,
    ATTAINING_TARGET,
    REACHED_TARGET,
    STOPPED,
    ERROR,
    SLOW_DOWN
};

class Module : public rclcpp::Node
{
public:
    Module(const std::string &name) : Node(name)
    {
        pivot_error_ = 0.0;
        actual_wheel_speed_ = 0.0;
        rqst_wheel_speed_ = 0.0;
        drive_wheel_position_ = 0.0;
        target_wheel_position_ = 0.0;
        rqst_pivot_angle_ = 0.0;
        actual_pivot_angle_ = 0.0;
        commanded_wheel_speed_ = 0.0;
        commanded_pivot_position_ = 0.0;
        limit_switch_triggered_ = false;
        show_changing_values_ = false;

        warning_flag_for_dime_ = false;

        module_name_ = this->declare_parameter<std::string>("module_name", "module_default");
        mock_encoder_values_ = this->declare_parameter<bool>("mock_encoder_values", true);
        invert_drive_motor_ = this->declare_parameter<bool>("invert_drive_motor", false);
        invert_pivot_motor_ = this->declare_parameter<bool>("invert_pivot_motor", false);
        odom_frequency_ = this->declare_parameter<double>("odom_frequency", 10.0);
        enable_logging_ = this->declare_parameter<bool>("enable_logging", false);

        if (invert_drive_motor_)
        {
            RCLCPP_INFO(this->get_logger(), "\033[32mInverting Drive Motor for %s\033[39m", module_name_.c_str());
        }

        command_sub_ = this->create_subscription<geometry_msgs::msg::Twist>("/" + module_name_ + "/command", 10, std::bind(&Module::command_callback, this, _1));
        encoder_pub_ = this->create_publisher<sensor_msgs::msg::JointState>("/" + module_name_ + "/encoders", 10);
        limit_switch_pub_ = this->create_publisher<std_msgs::msg::Float64>("/" + module_name_ + "/limit_switch", 10);
        pivot_position_pub_ = this->create_publisher<std_msgs::msg::Float64>("/" + module_name_ + "/pivot_position", 10);
        drive_motor_pub_ = this->create_publisher<std_msgs::msg::Float64>("/" + module_name_ + "/wheel_position", 10);
        speed_motor_pub_ = this->create_publisher<std_msgs::msg::Float64>("/" + module_name_ + "/wheel_speed", 10);
        pivot_error_pub_ = this->create_publisher<std_msgs::msg::Float64>("/" + module_name_ + "/pivot_error", 10);

        error_timer_ = this->create_wall_timer(std::chrono::milliseconds(500), std::bind(&Module::compute_pivot_error, this));
        update_timer_ = this->create_wall_timer(std::chrono::milliseconds(static_cast<int>(1000.0 / odom_frequency_)), std::bind(&Module::update_position, this));

        drive_pid_controller_ = std::make_unique<PIDController>(0.1, 0.4, 0.01, rqst_wheel_speed_, -2.0, 2.0, this->get_logger());
        pivot_pid_controller_ = std::make_unique<PIDController>(0.1, 0.1, 0.01, rqst_pivot_angle_, -6.28, 6.28, this->get_logger());

        RCLCPP_INFO(this->get_logger(), "\033[34m Module Ready: %s\033[39m", module_name_.c_str());
        set_state(ModuleState::IDLE);
    }

private:
    std::string module_name_;
    bool mock_encoder_values_;
    bool invert_drive_motor_;
    bool invert_pivot_motor_;
    bool warning_flag_for_dime_;
    bool limit_switch_triggered_;
    bool show_changing_values_;

    double odom_frequency_;
    double pivot_error_;
    double actual_wheel_speed_;
    double rqst_wheel_speed_;
    double drive_wheel_position_;
    double target_wheel_position_;
    double rqst_pivot_angle_;
    double actual_pivot_angle_;
    double commanded_wheel_speed_;
    double commanded_pivot_position_;

    rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr encoder_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr limit_switch_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr pivot_position_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr drive_motor_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr speed_motor_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr pivot_error_pub_;
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr command_sub_;

    rclcpp::TimerBase::SharedPtr error_timer_;
    rclcpp::TimerBase::SharedPtr update_timer_;

    std::unique_ptr<PIDController> drive_pid_controller_;
    std::unique_ptr<PIDController> pivot_pid_controller_;

    ModuleState module_state_;

    bool enable_logging_;

    void set_state(ModuleState state)
    {
        module_state_ = state;
        debugging_logging("\033[32mModule " + module_name_ + " state changed to: " + state_to_string(state) + "\033[39m");
    }

    std::string state_to_string(ModuleState state)
    {
        switch (state)
        {
        case ModuleState::IDLE:
            return "IDLE";
        case ModuleState::ATTAINING_TARGET:
            return "ATTAINING_TARGET";
        case ModuleState::STOPPED:
            return "STOPPED";
        case ModuleState::ERROR:
            return "ERROR";
        case ModuleState::REACHED_TARGET:
            return "REACHED_TARGET";
        case ModuleState::SLOW_DOWN:
            return "SLOW_DOWN";
        default:
            return "UNKNOWN";
        }
    }

    void command_callback(const geometry_msgs::msg::Twist::SharedPtr msg)
    {
        switch (module_state_)
        {
        case ModuleState::IDLE:
            set_rqst_pivot_angle(msg->angular.z);
            compute_pivot_error();
            set_rqst_wheel_speed(msg->linear.x);
            // print out the target wheel speed and pivot angle
            debugging_logging("Target wheel speed: " + std::to_string(msg->linear.x) + ", Target pivot angle: " + std::to_string(msg->angular.z));
            if (enable_logging_)
            {
                show_changing_values_ = true;
            }
            set_state(ModuleState::ATTAINING_TARGET);
            break;

        case ModuleState::ATTAINING_TARGET:
            if (std::abs(pivot_error_) < 0.1 && std::abs(rqst_wheel_speed_ - actual_wheel_speed_) < 0.1)
            {
                show_changing_values_ = false;
                set_state(ModuleState::REACHED_TARGET);
                debugging_logging("Reached target wheel speed: " + std::to_string(msg->linear.x) + ", target pivot angle: " + std::to_string(msg->angular.z));
            }

            // if the pivot error is too large, slow down the robot
            // if (std::abs(pivot_error_) > 0.5)
            // {
            //     set_state(ModuleState::SLOW_DOWN);
            //     debugging_logging("Pivot error too large, slowing down the robot.");
            // }

            break;

        case ModuleState::REACHED_TARGET:
            compute_pivot_error();
            
            if (msg->linear.x == 0.0)
            {
                set_state(ModuleState::STOPPED);
            }

            // track if the commanded value has changed
            if (msg->linear.x != rqst_wheel_speed_ || msg->angular.z != rqst_pivot_angle_)
            {

                // move me to the ATTAINING_TARGET state
                set_rqst_pivot_angle(msg->angular.z);
                compute_pivot_error();
                set_rqst_wheel_speed(msg->linear.x);
                set_state(ModuleState::ATTAINING_TARGET);
                debugging_logging("Target wheel speed: " + std::to_string(msg->linear.x) + ", Target pivot angle: " + std::to_string(msg->angular.z));
                show_changing_values_ = true;
            }

            break;

        case ModuleState::STOPPED:
            // print out some data for debugging
            debugging_logging("Module is stopped. Drive wheel speed: " + std::to_string(actual_wheel_speed_) + ", Pivot angle: " + std::to_string(actual_pivot_angle_));
            // print out some things about the message
            debugging_logging("Requested wheel speed: " + std::to_string(msg->linear.x) + ", Requested pivot angle: " + std::to_string(msg->angular.z));

            if (msg->linear.x != 0.0 || msg->angular.z != rqst_pivot_angle_)
            {
                set_state(ModuleState::IDLE);
            }
            break;

        case ModuleState::ERROR:
            debugging_logging("Module is in error state.");
            break;

        default:
            debugging_logging("Module is in an unknown state.");
            break;
        }
    }

    void debugging_logging(std::string message)
    {
        if (enable_logging_)
        {
            RCLCPP_INFO(this->get_logger(), "%s", message.c_str());
        }
    }

    void compute_pivot_error()
    {
        pivot_error_ = rqst_pivot_angle_ - actual_pivot_angle_;
        std_msgs::msg::Float64 error_msg;
        error_msg.data = pivot_error_;
        pivot_error_pub_->publish(error_msg);
    }

    void set_rqst_wheel_speed(double speed)
    {
        rqst_wheel_speed_ = speed;
    }

    void set_rqst_pivot_angle(double angle)
    {
        rqst_pivot_angle_ = angle;
    }

    void update_encoder_values()
    {
        if (!mock_encoder_values_)
        {
            throw std::runtime_error("Actual encoder values not implemented yet");
        }
        else
        {
            actual_wheel_speed_ = commanded_wheel_speed_;
            actual_pivot_angle_ = commanded_pivot_position_;
        }
    }

    void update_limit_switch()
    {
        if (!mock_encoder_values_)
        {
            throw std::runtime_error("Actual limit switch state not implemented yet");
        }
        else
        {
            limit_switch_triggered_ = false;
        }
    }

    void update_position()
    {
        update_encoder_values();
        update_limit_switch();

        // -------------------------- Linear Motion --------------------------------------
        double drive_output = 0.0;

        if (std::abs(pivot_error_) < 0.1)
        {
            drive_output = drive_pid_controller_->compute_move(rqst_wheel_speed_, actual_wheel_speed_);
        }
        else
        {
            drive_output = drive_pid_controller_->compute_move( 0 , actual_wheel_speed_);
        }

        // Publish the speed and drive motor values
        std_msgs::msg::Float64 speed_msg;
        speed_msg.data = invert_drive_motor_ ? -drive_output : drive_output;
        speed_motor_pub_->publish(speed_msg);

        drive_wheel_position_ += invert_drive_motor_ ? -drive_output : drive_output;
        std_msgs::msg::Float64 drive_msg;
        drive_msg.data = drive_wheel_position_;
        drive_motor_pub_->publish(drive_msg);

        // -------------------------- Pivot Motion --------------------------------------

        // if the speed error is too high and the pivot error is too high, don't change the pivot angle
        double pivot_output = actual_pivot_angle_; // default to the current pivot angle
        if (!(std::abs(actual_wheel_speed_) > 0.1 && std::abs(pivot_error_) > 0.1))
        {
            // if we're within the limit switch range, move the pivot
            pivot_output = pivot_pid_controller_->compute_move(rqst_pivot_angle_, actual_pivot_angle_);
        }

        std_msgs::msg::Float64 pivot_msg;
        pivot_msg.data = invert_pivot_motor_ ? -pivot_output : pivot_output;
        pivot_position_pub_->publish(pivot_msg);

        // print out the drive wheel position and wheel angle
        if (show_changing_values_)
        {
           debugging_logging("Drive wheel position: " + std::to_string(drive_wheel_position_) + ", Pivot angle: " + std::to_string(actual_pivot_angle_));
        }

        if (mock_encoder_values_)
        {
            commanded_wheel_speed_ = drive_output;
            commanded_pivot_position_ = pivot_output;
        }
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<Module>("hermes_swerve_module");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
